File: bot/src/service/StabilityAiService.ts
import {Context} from "../Context";
import {Logger} from "./LoggingService";
import axios, {AxiosError} from "axios";
import ServiceError from "../ServiceError";
import FormData from "form-data";

export type AspectRatio = "16:9" | "1:1" | "21:9" | "2:3" | "3:2" | "4:5" | "5:4" | "9:16" | "9:21";

export default class StabilityAiService {
    private logger!: Logger;
    private apiKey!: string;

    constructor(private context: Context) {
        this.context.onReady(() => {
            this.logger = context.loggingService.getRootLogger().newSublogger('StabilityAiService');
            this.apiKey = context.configService.getAppConfig().stabilityAiApiKey;
        });
    }

    async generateImage(
        prompt: string,
        negativePrompt: string | null = null,
        aspectRatio: AspectRatio = "1:1"
    ): Promise<Buffer> {
        const url = "https://api.stability.ai/v2beta/stable-image/generate/ultra";
        const form = new FormData();
        form.append('prompt', prompt);
        if (negativePrompt) {
            form.append('negative_prompt', negativePrompt);
        }
        form.append('aspect_ratio', aspectRatio);
        form.append('output_format', "png");
        try {
            const response = await axios.post(url, form, this.createConfig());
            return Buffer.from(response.data);
        } catch (e: unknown) {
            this.logger.error("Failed to generate image");
            console.error(e);
            throw this.createStabilityAiWrapperError(e);
        }
    }

    async searchAndReplace(
        image: Buffer,
        prompt: string,
        searchPrompt: string,
        negativePrompt: string | null = null
    ): Promise<Buffer> {
        const url = "https://api.stability.ai/v2beta/stable-image/edit/search-and-replace";
        const form = new FormData();
        form.append('image', image, {
            filename: 'image.jpg',
            contentType: 'image/jpeg',
        });
        form.append('prompt', prompt);
        form.append('search_prompt', searchPrompt);
        if (negativePrompt) {
            form.append('negative_prompt', negativePrompt);
        }
        form.append('output_format', "png");
        try {
            const response = await axios.post(url, form, this.createConfig());
            return Buffer.from(response.data);
        } catch (e: unknown) {
            this.logger.error("Failed to search and replace image");
            console.error(e);
            throw this.createStabilityAiWrapperError(e);
        }
    }

    async removeBackground(image: Buffer): Promise<Buffer> {
        const url = "https://api.stability.ai/v2beta/stable-image/edit/remove-background";
        const form = new FormData();
        form.append('image', image, {
            filename: 'image.jpg',
            contentType: 'image/jpeg',
        });
        form.append('output_format', "png");
        try {
            const response = await axios.post(url, form, this.createConfig());
            return Buffer.from(response.data);
        } catch (e: unknown) {
            this.logger.error("Failed to remove image background");
            console.error(e);
            throw this.createStabilityAiWrapperError(e);
        }
    }

    async sketch(
        image: Buffer,
        prompt: string,
        controlStrength: number = 0.7,
        negativePrompt: string | null = null
    ): Promise<Buffer> {
        const url = "https://api.stability.ai/v2beta/stable-image/control/sketch";
        const form = new FormData();
        form.append('image', image, {
            filename: 'image.jpg',
            contentType: 'image/jpeg',
        });
        form.append('prompt', prompt);
        form.append('control_strength', controlStrength.toString());
        if (negativePrompt) {
            form.append('negative_prompt', negativePrompt);
        }
        form.append('output_format', "png");
        try {
            const response = await axios.post(url, form, this.createConfig());
            return Buffer.from(response.data);
        } catch (e: unknown) {
            this.logger.error("Failed to draw from sketch");
            console.error(e);
            throw this.createStabilityAiWrapperError(e);
        }
    }

    async structure(
        image: Buffer,
        prompt: string,
        controlStrength: number = 0.7,
        negativePrompt: string | null = null
    ): Promise<Buffer> {
        const url = "https://api.stability.ai/v2beta/stable-image/control/structure";
        const form = new FormData();
        form.append('image', image, {
            filename: 'image.jpg',
            contentType: 'image/jpeg',
        });
        form.append('prompt', prompt);
        form.append('control_strength', controlStrength.toString());
        if (negativePrompt) {
            form.append('negative_prompt', negativePrompt);
        }
        form.append('output_format', "png");
        try {
            const response = await axios.post(url, form, this.createConfig());
            return Buffer.from(response.data);
        } catch (e: unknown) {
            this.logger.error("Failed to control structure");
            console.error(e);
            throw this.createStabilityAiWrapperError(e);
        }
    }

    private createConfig(): object {
        return {
            responseType: 'arraybuffer',
            headers: {
                'Authorization': `Bearer ${this.apiKey}`,
                'Content-Type': "multipart/form-data",
                'Accept': "image/*"
            }
        }
    }

    private createStabilityAiWrapperError(e: unknown): Error {
        if (axios.isAxiosError(e)) {
            const axiosError = e as AxiosError;
            if (axiosError.response && axiosError.response.data) {
                const data = axiosError.response.data as any;
                const message = Array.isArray(data.errors) ? data.errors.join(", ") : axiosError.message;
                return new ServiceError("StabilityAI: " + message)
            } else {
                return new ServiceError("API call failed: " + axiosError.message);
            }
        } else if (e instanceof Error) {
            const error = e as Error
            return new ServiceError("Service call failed: " + error.message)
        }
        return new ServiceError("Unknown problem, please check logs")
    }
}

File: backend/src/main/java/space/davids_digital/sweetie/integration/stabilityai/StabilityAiService.kt
package space.davids_digital.sweetie.integration.stabilityai

import okhttp3.*
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.RequestBody.Companion.toRequestBody
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Service
import java.io.IOException

@Service
class StabilityAiService(
    @Qualifier("stabilityAiApiKey")
    private val apiKey: String,
) {
    companion object {
        private val log = LoggerFactory.getLogger(StabilityAiService::class.java)
        private val ASPECT_RATIOS = listOf("16:9", "1:1", "21:9", "2:3", "3:2", "4:5", "5:4", "9:16", "9:21")
        private val IMAGE_MEDIA_TYPE = "image/jpeg".toMediaTypeOrNull()
    }

    private val client = OkHttpClient.Builder()
        .connectTimeout(60, java.util.concurrent.TimeUnit.SECONDS)
        .writeTimeout(120, java.util.concurrent.TimeUnit.SECONDS)
        .readTimeout(300, java.util.concurrent.TimeUnit.SECONDS)
        .build()

    fun generate(prompt: String, negativePrompt: String? = null, aspectRatio: String = "1:1"): ByteArray {
        validateAspectRatio(aspectRatio)
        val params = mutableMapOf(
            "prompt" to prompt,
            "aspect_ratio" to aspectRatio,
            "output_format" to "png"
        )
        negativePrompt?.let { params["negative_prompt"] = it }
        return call("https://api.stability.ai/v2beta/stable-image/generate/ultra", params)
    }

    fun searchAndReplace(
        image: ByteArray,
        prompt: String,
        searchPrompt: String,
        negativePrompt: String? = null,
    ): ByteArray {
        val params = mutableMapOf(
            "prompt" to prompt,
            "search_prompt" to searchPrompt,
            "output_format" to "png"
        )
        negativePrompt?.let { params["negative_prompt"] = it }
        return call("https://api.stability.ai/v2beta/stable-image/edit/search-and-replace", params, image)
    }

    fun removeBackground(image: ByteArray): ByteArray {
        val params = mapOf("output_format" to "png")
        return call("https://api.stability.ai/v2beta/stable-image/edit/remove-background", params, image)
    }

    fun sketch(
        image: ByteArray,
        prompt: String,
        controlStrength: Double = 0.7,
        negativePrompt: String? = null,
    ): ByteArray {
        val params = mutableMapOf(
            "prompt" to prompt,
            "control_strength" to controlStrength.toString(),
            "output_format" to "png"
        )
        negativePrompt?.let { params["negative_prompt"] = it }
        return call("https://api.stability.ai/v2beta/stable-image/control/sketch", params, image)
    }

    fun structure(
        image: ByteArray,
        prompt: String,
        controlStrength: Double = 0.7,
        negativePrompt: String? = null,
    ): ByteArray {
        val params = mutableMapOf(
            "prompt" to prompt,
            "control_strength" to controlStrength.toString(),
            "output_format" to "png"
        )
        negativePrompt?.let { params["negative_prompt"] = it }
        return call("https://api.stability.ai/v2beta/stable-image/control/structure", params, image)
    }

    private fun validateAspectRatio(aspectRatio: String) {
        if (aspectRatio !in ASPECT_RATIOS) {
            throw IllegalArgumentException("Aspect ratio must be one of $ASPECT_RATIOS, but was $aspectRatio")
        }
    }

    private fun call(
        url: String,
        params: Map<String, String>,
        image: ByteArray? = null
    ): ByteArray {
        val request = buildRequest(url, params, image)
        return executeRequest(request)
    }

    private fun buildRequest(
        url: String,
        params: Map<String, String>,
        image: ByteArray? = null
    ): Request {
        val formBodyBuilder = MultipartBody.Builder().setType(MultipartBody.FORM)
        for ((key, value) in params) {
            formBodyBuilder.addFormDataPart(key, value)
        }
        image?.let {
            formBodyBuilder.addFormDataPart(
                "image", "image.jpg",
                it.toRequestBody(IMAGE_MEDIA_TYPE)
            )
        }
        val requestBody = formBodyBuilder.build()
        return Request.Builder()
            .url(url)
            .post(requestBody)
            .addHeader("Authorization", "Bearer $apiKey")
            .addHeader("Accept", "image/*")
            .build()
    }

    private fun executeRequest(request: Request): ByteArray {
        client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                val errorBody = response.body?.string()
                log.error("Request failed: ${response.code} - $errorBody")
                throw RuntimeException("StabilityAI: ${response.message} - $errorBody")
            }
            return response.body?.bytes() ?: throw IOException("Empty response body")
        }
    }
}
